File: aoc/2023/aoc202308.hs
import Data.Map(Map,fromList,(!))
import Data.Tuple(swap)

parse :: String -> ([(String,(String,String))],String)
parse = fmap concat . swap . fmap (map parseEdge) . splitAt 2 . lines
  where
    parseEdge = formatEdge . words . filter (not . (`elem` "=(,)"))
    formatEdge [a,b,c] = (a,(b,c))

countSteps :: Map String (String,String) -> (String -> Bool) -> Int -> String -> String -> Int
countSteps graph goal n instructions current
  | goal current = n
  | head instructions == 'L' =
      countSteps graph goal (n+1) (tail instructions) (fst (graph!current))
  | head instructions == 'R' =
      countSteps graph goal (n+1) (tail instructions) (snd (graph!current))

result :: String -> Int
result input = countSteps (fromList graph) ("ZZZ" ==) 0 (cycle instructions) "AAA"
  where (graph,instructions) = parse input

testData :: [String]
testData = [unlines [
    "RL",
    "",
    "AAA = (BBB, CCC)",
    "BBB = (DDD, EEE)",
    "CCC = (ZZZ, GGG)",
    "DDD = (DDD, DDD)",
    "EEE = (EEE, EEE)",
    "GGG = (GGG, GGG)",
    "ZZZ = (ZZZ, ZZZ)"
    ], unlines [
    "LLR",
    "",
    "AAA = (BBB, BBB)",
    "BBB = (AAA, ZZZ)",
    "ZZZ = (ZZZ, ZZZ)"
    ]]

test :: ()
test
  | map result testData /= [2,6] = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap result $ readFile "input/08.txt"

result2 :: String -> Int
result2 input = foldl lcm 1 counts
  where
    (graph,instructions) = parse input
    counts = map (countSteps (fromList graph) ((== 'Z') . last) 0 (cycle instructions)) (filter ((== 'A') . last) (map fst graph))

testData2 :: String
testData2 = unlines [
    "LR",
    "",
    "11A = (11B, XXX)",
    "11B = (XXX, 11Z)",
    "11Z = (11B, XXX)",
    "22A = (22B, XXX)",
    "22B = (22C, 22C)",
    "22C = (22Z, 22Z)",
    "22Z = (22B, 22B)",
    "XXX = (XXX, XXX)"
    ]

test2 :: ()
test2
  | result2 testData2 /= 6 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap result2 $ readFile "input/08.txt"

File: aoc/2025/aoc202308.go
package main

func init() {
	Register(&aoc202308{
		AOC: AOC{
			Day:           8,
			InputFilename: "../2023/input/08.txt",
			Tests: []Test{
				Test{
					Input: `RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)
`,
					Part1: "2",
					Part2: "",
				},
				Test{
					Input: `LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
`,
					Part1: "6",
					Part2: "",
				},
				Test{
					Input: `LR

11A = (11B, XXX)
11B = (XXX, 11Z)
11Z = (11B, XXX)
22A = (22B, XXX)
22B = (22C, 22C)
22C = (22Z, 22Z)
22Z = (22B, 22B)
XXX = (XXX, XXX)
`,
					Part1: "",
					Part2: "6",
				},
			},
		},
	})
}

type aoc202308 struct {
	AOC
}

func (aoc *aoc202308) parse(input *Input) (string, map[string][2]string) {
	instructions, _ := input.Word()
	graph := map[string][2]string{}
	for {
		w, ok := input.Word()
		if !ok {
			break
		}
		input.Word()
		l, _ := input.Word()
		r, _ := input.Word()
		graph[w] = [2]string{l[1 : len(l)-1], r[:len(r)-1]}
	}
	return instructions, graph
}

func (aoc *aoc202308) Part1(input *Input) string {
	instructions, graph := aoc.parse(input)
	steps := 0
	node := "AAA"
	dirs := instructions
	for node != "ZZZ" {
		if dirs == "" {
			dirs = instructions
		}
		dir := 0
		if dirs[0] == 'R' {
			dir = 1
		}
		dirs = dirs[1:]
		node = graph[node][dir]
		steps++
	}
	return IntResult(steps)
}

func (aoc *aoc202308) Part2(input *Input) string {
	instructions, graph := aoc.parse(input)
	result := 1
	for node, _ := range graph {
		// In my data, the end is reached after an integral number
		// of times through the instructions in every path, and
		// the end then goes to the same node as the start does.
		if node[len(node)-1] == 'A' {
			steps := 0
			n := node
			dirs := instructions
			for n[len(n)-1] != 'Z' {
				if dirs == "" {
					dirs = instructions
				}
				dir := 0
				if dirs[0] == 'R' {
					dir = 1
				}
				dirs = dirs[1:]
				n = graph[n][dir]
				steps++
			}
			result = LCM(result, steps)
		}
	}
	return IntResult(result)
}
